clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -242,46 +317,36 @@ public PrototypeAsyncConsumer(LogContext logContext,
      */
     @Override
     public ConsumerRecords<K, V> poll(final Duration timeout) {
+        Timer timer = time.timer(timeout);
+
         try {
-            Timer timer = time.timer(timeout);
+            this.kafkaConsumerMetrics.recordPollStart(timer.currentTimeMs());
 
-            do {
-                if (!eventHandler.isEmpty()) {
-                    final Optional<BackgroundEvent> backgroundEvent = eventHandler.poll();
-                    // processEvent() may process 3 types of event:
-                    // 1. Errors
-                    // 2. Callback Invocation
-                    // 3. Fetch responses
-                    // Errors will be handled or rethrown.
-                    // Callback invocation will trigger callback function execution, which is blocking until completion.
-                    // Successful fetch responses will be added to the completedFetches in the fetcher, which will then
-                    // be processed in the collectFetches().
-                    backgroundEvent.ifPresent(event -> log.warn("Do something with this background event: {}", event));
-                }
+            if (this.subscriptions.hasNoSubscriptionOrUserAssignment()) {
+                throw new IllegalStateException("Consumer is not subscribed to any topics or assigned any partitions");
+            }
 
-                updateAssignmentMetadataIfNeeded();
+            do {
+                updateAssignmentMetadataIfNeeded(timer);
+                final Fetch<K, V> fetch = pollForFetches(timer);
```

Comment:
Maybe it is out of the scope of this PR but we also want to drain the BackgroundEventQueue to process the events.

Reply:
I've added a new class (`BackgroundEventProcessor`) which is called by the `PrototypeAsyncConsumer` to perform that logic.